brush-core/src/builtins/ulimit.rs

```diff
@@ -0,0 +1,489 @@
+        help: "the maximum amount of cpu time in seconds",
+        description: "cpu time",
+        short: 't',
+        unit: Unit::Seconds,
+    };
+    const NPROC: ResourceDescription = ResourceDescription {
+        resource: Resource::Phy(rlimit::Resource::NPROC),
+        help: "the maximum number of user processes",
+        description: "max user processes",
+        short: 'u',
+        unit: Unit::Number,
+    };
+    const VMEM: ResourceDescription = ResourceDescription {
+        resource: Resource::Virt(Virtual::VMem),
+        help: "the size of virtual memory",
+        description: "virtual memory",
+        short: 'v',
+        unit: Unit::KBytes,
+    };
+    const THREADS: ResourceDescription = ResourceDescription {
+        resource: Resource::Phy(rlimit::Resource::THREADS),
+        help: "the maximum number of threads",
+        description: "number of threads",
+        short: 'T',
+        unit: Unit::Number,
+    };
+    const NPTS: ResourceDescription = ResourceDescription {
+        resource: Resource::Phy(rlimit::Resource::NPTS),
+        help: "the maximum number of pseudoterminals",
+        description: "number of pseudoterminals",
+        short: 'P',
+        unit: Unit::Number,
+    };
+
+    fn get(&self, hard: bool) -> std::io::Result<String> {
+        let (soft_limit, hard_limit) = self.resource.get()?;
+        let val = if hard { hard_limit } else { soft_limit };
+
+        if val == rlimit::INFINITY {
+            Ok("unlimited".into())
+        } else {
+            Ok(format!("{}", val / self.unit.scale()))
+        }
+    }
+
+    fn set(&self, set_hard: bool, value: LimitValue) -> std::io::Result<()> {
+        let (soft, hard) = self.resource.get()?;
+        let value = match value {
+            LimitValue::Soft => soft,
+            LimitValue::Hard => hard,
+            LimitValue::Unlimited => rlimit::INFINITY,
+            LimitValue::Value(v) => v * self.unit.scale(),
+            LimitValue::Unset => return Ok(()),
+        };
+
+        if set_hard {
+            self.resource.set(soft, value)
+        } else {
+            self.resource.set(value, hard)
+        }
+    }
+
+    /// Print either soft or hard limit
+    fn print(&self, hard: bool) {
+        if !self.resource.is_supported() {
+            return;
+        }
+        let unit = match self.unit {
+            Unit::Block => format!("(block, -{})", self.short),
+            Unit::Bytes => format!("(bytes, -{})", self.short),
+            Unit::HalfKBytes => format!("(512 bytes, -{})", self.short),
+            Unit::KBytes => format!("(kbytes, -{})", self.short),
+            Unit::Micros => format!("(microseconds, -{})", self.short),
+            Unit::Number => format!("(-{})", self.short),
+            Unit::Seconds => format!("(seconds, -{})", self.short),
+        };
+        let resource = self.get(hard).unwrap_or_else(|e| format!("{e}"));
+        println!("{:<26}{:>16} {}", self.description, unit, resource);
```

Comment:
I forgot to note this earlier -- instead of calling `print*`, builtins should use `write*` with `context.stdout()` or `context.stderr()`. That ensures it will go to the right output stream for the current execution context (in case it was redirected).

Ditto for the other calls to `print*` in this file.